src/protocols/secure_channel/CASESession.cpp

```diff
@@ -2224,6 +2336,18 @@ CHIP_ERROR CASESession::OnMessageReceived(ExchangeContext * ec, const PayloadHea
     // Call delegate to indicate session establishment failure.
     if (err != CHIP_NO_ERROR)
     {
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+        // Release the ExchangeContext that was retained upon receipt of Sigma1.
+        // The Retain() was necessary to prevent the automatic release of the
+        // ExchangeContext after sending the last message of the Exchange and
+        // before the call to Finish()
+        if (mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->GetPeerAddress().GetTransportType() ==
+            Transport::Type::kTcp)
+        {
+            mExchangeCtxt->Release();
```

Comment:
Again, we can be here without having done the retain.

Reply:
Changed to ExchangeHandle.